sql/src/main/java/io/crate/planner/consumer/MultiSourceGroupByConsumer.java

```diff
@@ -87,39 +90,115 @@ public Plan visitMultiSourceSelect(MultiSourceSelect multiSourceSelect, Consumer
             }
 
             removePostGroupingActionsFromQuerySpec(multiSourceSelect, splitPoints);
-            Plan plan = createSubPlan(context, multiSourceSelect);
-            addGroupProtection(plan, splitPoints, groupKeys);
-            addFilterProtection(plan, groupKeys, splitPoints, querySpec);
-            addTopN(plan, context, splitPoints, groupKeys, querySpec, outputs);
+
+            Plan plan = context.plannerContext().planSubRelation(multiSourceSelect, context);
+
+            if (isExecutedOnHandler(context, plan)) {
+                addNonDistributedGroupProjection(plan, splitPoints, groupKeys);
+                addFilterProjection(plan, splitPoints, groupKeys, querySpec);
+                addTopN(plan, splitPoints, groupKeys, querySpec, context, outputs);
+            } else {
+                addDistributedGroupProjection(plan, splitPoints, groupKeys);
+                plan = createReduceMerge(plan, splitPoints, groupKeys, querySpec, context, outputs);
+            }
 
             return plan;
-        };
+        }
 
         /**
-         * Create and return the execution plan depending on distributed or non distributed execution.
+         * Returns true for non distributed and false for distributed execution.
          */
-        private static Plan createSubPlan(ConsumerContext context, MultiSourceSelect multiSourceSelect) {
-            Plan plan = context.plannerContext().planSubRelation(multiSourceSelect, context);
-            boolean executesOnHandler = ExecutionPhases.executesOnHandler(
+        private static boolean isExecutedOnHandler(ConsumerContext context, Plan plan) {
+            return ExecutionPhases.executesOnHandler(
                 context.plannerContext().handlerNode(),
                 plan.resultDescription().nodeIds()
             );
-            if (!executesOnHandler) {
-                // TODO: support distributed execution.
-                throw new UnsupportedFeatureException("Distributed execution is not supported");
-            }
+        }
 
-            return plan;
+        /**
+         * Creates and returns the merge plan.
+         */
+        private Merge createReduceMerge(Plan plan,
+                                        SplitPoints splitPoints,
+                                        List<Symbol> groupKeys,
+                                        QuerySpec querySpec,
+                                        ConsumerContext context,
+                                        List<Symbol> outputs) {
+            ArrayList<Symbol> groupProjectionOutputs = new ArrayList<>(groupKeys);
+            groupProjectionOutputs.addAll(splitPoints.aggregates());
+
+            ProjectionBuilder projectionBuilder = new ProjectionBuilder(functions);
```

Comment:
maybe we could create the ProjectionBuilder in the constructor. It doesn't contain state and I think now there are paths where it would be created multiple times.